features/include/pcl/features/range_image_border_extractor.h

```diff
@@ -159,36 +159,35 @@ namespace pcl
       getBorderScoresBottom () { extractBorderScoreImages (); return border_scores_bottom_.data (); }
 
       LocalSurface**
-      getSurfaceStructure () { extractLocalSurfaceStructure (); return surface_structure_; }
+      getSurfaceStructure () { extractLocalSurfaceStructure (); return surface_structure_.get(); }
```

Comment:
Is it a good idea to return the double pointer here, since the ownership is not transferred here?

Reply:
Yes. C++ best practices: raw pointers are non-owning. In this class, the actual delete is done by the dtor of the class (implying the class is the owner)

Reply:
Couldn't this lead to a dangling pointer when someone accesses the pointer after the classes destructor has been called?

Reply:
That's part of the contract of not having ownership. The only way to circumvent it would be with a weak_ptr, but that will imply shared ownership because you can't access the underlying object with a weak_ptr.